src/compiler/types.ts

```diff
@@ -4238,6 +4238,13 @@ namespace ts {
         reportsUnnecessary?: {};
         code: number;
         source?: string;
+        relatedInformation?: DiagnosticRelatedInformation[];
+    }
+    export interface DiagnosticRelatedInformation {
+        file: SourceFile | undefined;
+        start: number | undefined;
+        length: number | undefined;
+        messageText: string | DiagnosticMessageChain;

```

Comment:
>messageText: string | DiagnosticMessageChain; [](start = 8, length = 45)

I think I need more information about how these are intended to be used.  Roslyn diagnostics support additional locations, but those locations don't have messages.  We could create separate diagnostics but we would need a new code and category and I don't believe we'd be able to link them together.  I think Roslyn more or less dropped this functionality (vs the old native compiler).

Reply:
Secondary code locations and more information pertaining to specifically that location in the context of the diagnostic. (eg, "Defined here", "Declared here", "Replicated here", and so on. The additional spans also allow us to use those spans when calculating quick fixes - so we can, for example, know what we can suggest a cast at the site of an assignability error, but also suggest modifying the type the assignment is failing against (because it is mentioned as a related span). The only diagnostic set we currently do a similar thing with is the `esModuleInterop` import error, which we currently issue a secondary diagnostic for (so there's no obvious link between the two).

Reply:
For reference, this isn't a new concept and there is prior art, which is why vscode has basic support for them (and is looking to iterate on their UI) - [rust makes good use of related spans in it's error](ttps://blog.rust-lang.org/2016/08/10/Shape-of-errors-to-come.html).

Reply:
We need to sit down and design this feature.. there seems to be two uses cases, 
1. additional locations, e.g. duplicate identifier errors, you would list them all in the same error message
2. related diagnostics, e.g. type of var `x` is not assignable to string, and here is where `x` was declated